File: nano.py
#!/usr/bin/env python3
"""nano-opencode: 400 lines to beat them all."""

import json, subprocess, sys, signal, time
from pathlib import Path

try:
    import litellm
except ImportError:
    print("Run: pip install litellm")
    sys.exit(1)

# ══════════════════════════════════════════════════════════════════════════════
# CONSTANTS
# ══════════════════════════════════════════════════════════════════════════════

MODEL = "anthropic/claude-sonnet-4-20250514"
TOKEN_LIMIT = 100_000
TOOL_LIMIT = 50
TIMEOUT = 120
TRUNCATE_CHARS = 30_000

# ══════════════════════════════════════════════════════════════════════════════
# SYSTEM PROMPT
# ══════════════════════════════════════════════════════════════════════════════

SYSTEM = """You are Nano, an expert software engineer operating in a terminal.

## Tools
You have 6 tools: shell, patch, read, write, find, grep.

## Workflow
1. EXPLORE - Understand the codebase, find relevant files
2. ANALYZE - Read code to understand patterns and style
3. PLAN - Design minimal, precise changes
4. EXECUTE - Apply changes using patch (edits) or write (new files)
5. VERIFY - Run tests/linters to confirm changes work

## Rules
- State your intent before each tool call
- Keep changes minimal and atomic
- Match existing code style exactly
- Verify changes work before declaring success
- If a patch fails, re-read the file to get exact content

{context}"""

# ══════════════════════════════════════════════════════════════════════════════
# TOOL SCHEMAS
# ══════════════════════════════════════════════════════════════════════════════

TOOLS = [
    {"type": "function", "function": {
        "name": "shell",
        "description": "Run shell command. Use for: git, tests, builds, installing packages. Output truncated.",
        "parameters": {"type": "object", "properties": {
            "cmd": {"type": "string", "description": "Command to execute"}
        }, "required": ["cmd"]}
    }},
    {"type": "function", "function": {
        "name": "patch",
        "description": "Replace exact text in file. Search must match exactly once. If fails, use read first.",
        "parameters": {"type": "object", "properties": {
            "file": {"type": "string", "description": "File path"},
            "search": {"type": "string", "description": "Exact text to find (including whitespace)"},
            "replace": {"type": "string", "description": "Text to replace with"}
        }, "required": ["file", "search", "replace"]}
    }},
    {"type": "function", "function": {
        "name": "read",
        "description": "Read file contents with line numbers.",
        "parameters": {"type": "object", "properties": {
            "path": {"type": "string", "description": "File path"},
            "start": {"type": "integer", "description": "Start line (1-indexed)"},
            "end": {"type": "integer", "description": "End line"}
        }, "required": ["path"]}
    }},
    {"type": "function", "function": {
        "name": "write",
        "description": "Create or overwrite a file.",
        "parameters": {"type": "object", "properties": {
            "path": {"type": "string", "description": "File path"},
            "content": {"type": "string", "description": "File content"}
        }, "required": ["path", "content"]}
    }},
    {"type": "function", "function": {
        "name": "find",
        "description": "List files in directory with optional glob pattern.",
        "parameters": {"type": "object", "properties": {
            "path": {"type": "string", "description": "Directory (default: .)"},
            "pattern": {"type": "string", "description": "Glob pattern (e.g., **/*.py)"}
        }, "required": []}
    }},
    {"type": "function", "function": {
        "name": "grep",
        "description": "Search for pattern in files. Returns matching lines.",
        "parameters": {"type": "object", "properties": {
            "pattern": {"type": "string", "description": "Regex pattern"},
            "path": {"type": "string", "description": "Directory to search (default: .)"},
            "include": {"type": "string", "description": "File glob (e.g., *.py)"}
        }, "required": ["pattern"]}
    }}
]

# ══════════════════════════════════════════════════════════════════════════════
# TOOLS IMPLEMENTATION
# ══════════════════════════════════════════════════════════════════════════════

def truncate(text: str, max_chars: int = TRUNCATE_CHARS) -> str:
    """Truncate keeping head (25%) and tail (75%) - errors usually at end."""
    if len(text) <= max_chars:
        return text
    head = max_chars // 4
    tail = max_chars - head
    return f"{text[:head]}\n\n... [{len(text) - max_chars:,} chars truncated] ...\n\n{text[-tail:]}"


class Tools:
    def __init__(self, root: Path):
        self.root = root.resolve()

    def safe(self, path: str) -> Path:
        """Resolve path and prevent traversal outside repo."""
        p = (self.root / path).resolve()
        if not str(p).startswith(str(self.root)):
            raise ValueError(f"Path outside repo: {path}")
        return p

    def shell(self, cmd: str) -> str:
        try:
            r = subprocess.run(["bash", "-c", cmd], cwd=self.root, timeout=TIMEOUT,
                               capture_output=True, text=True)
            out = (r.stdout + r.stderr).strip()
            if not out:
                return "Command completed (no output)" if r.returncode == 0 else f"Failed (exit {r.returncode})"
            return truncate(out)
        except subprocess.TimeoutExpired:
            return f"Timeout after {TIMEOUT}s"
        except Exception as e:
            return f"Error: {e}"

    def patch(self, file: str, search: str, replace: str) -> str:
        try:
            p = self.safe(file)
            if not p.exists():
                return f"File not found: {file}"
            content = p.read_text()
            count = content.count(search)
            if count == 0:
                return "Search not found. Use 'read' to see exact content."
            if count > 1:
                return f"Found {count} matches. Add more context to make unique."
            p.write_text(content.replace(search, replace, 1))
            return "Patch applied"
        except Exception as e:
            return f"Error: {e}"

    def read(self, path: str, start: int = None, end: int = None) -> str:
        try:
            p = self.safe(path)
            if not p.exists():
                return f"File not found: {path}"
            lines = p.read_text().splitlines()
            if start or end:
                start = (start or 1) - 1
                end = end or len(lines)
                lines = lines[start:end]
                start_num = start + 1
            else:
                start_num = 1
            numbered = [f"{i+start_num:4d}| {l}" for i, l in enumerate(lines)]
            return truncate("\n".join(numbered))
        except Exception as e:
            return f"Error: {e}"

    def write(self, path: str, content: str) -> str:
        try:
            p = self.safe(path)
            p.parent.mkdir(parents=True, exist_ok=True)
            p.write_text(content)
            return f"Written: {path} ({len(content)} bytes)"
        except Exception as e:
            return f"Error: {e}"

    def find(self, path: str = ".", pattern: str = None) -> str:
        try:
            p = self.safe(path)
            files = list(p.glob(pattern)) if pattern else list(p.iterdir())
            result = []
            for f in sorted(files)[:100]:
                rel = f.relative_to(self.root)
                prefix = "D " if f.is_dir() else "F "
                result.append(f"{prefix}{rel}")
            if len(files) > 100:
                result.append(f"... and {len(files) - 100} more")
            return "\n".join(result) or "No files found"
        except Exception as e:
            return f"Error: {e}"

    def grep(self, pattern: str, path: str = ".", include: str = None) -> str:
        try:
            p = self.safe(path)
            cmd = ["grep", "-rn", "--include", include, pattern, str(p)] if include else \
                  ["grep", "-rn", pattern, str(p)]
            r = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
            return truncate(r.stdout) if r.stdout else "No matches"
        except Exception as e:
            return f"Error: {e}"

    def execute(self, name: str, args: dict) -> str:
        method = getattr(self, name, None)
        if not method:
            return f"Unknown tool: {name}"
        try:
            return method(**args)
        except Exception as e:
            return f"Tool error: {e}"

# ══════════════════════════════════════════════════════════════════════════════
# AGENT
# ══════════════════════════════════════════════════════════════════════════════

def load_context(root: Path) -> str:
    """Load AGENT.md or similar project context file."""
    for name in ["AGENT.md", ".agent.md", "CLAUDE.md", ".claude.md"]:
        p = root / name
        if p.exists():
            return f"\n## Project Context\n{p.read_text()}"
    return ""


class Agent:
    def __init__(self, root: Path, model: str = MODEL):
        self.root = root
        self.model = model
        self.tools = Tools(root)
        self.messages = []
        self.calls = 0
        self.tokens = 0

    def run(self, task: str) -> str:
        context = load_context(self.root)
        self.messages = [
            {"role": "system", "content": SYSTEM.format(context=context)},
            {"role": "user", "content": task}
        ]
        self.calls = 0

        while self.calls < TOOL_LIMIT:
            try:
                r = litellm.completion(model=self.model, messages=self.messages, 
                                       tools=TOOLS, tool_choice="auto")
            except Exception as e:
                return f"LLM error: {e}"

            msg = r.choices[0].message
            self.tokens += r.usage.total_tokens if r.usage else 0
            
            # Convert to dict for appending
            msg_dict = {"role": "assistant", "content": msg.content or ""}
            if msg.tool_calls:
                msg_dict["tool_calls"] = [
                    {"id": tc.id, "type": "function", 
                     "function": {"name": tc.function.name, "arguments": tc.function.arguments}}
                    for tc in msg.tool_calls
                ]
            self.messages.append(msg_dict)

            # Print thinking
            if msg.content:
                print(f"\n{msg.content}")

            if not msg.tool_calls:
                break

            for tc in msg.tool_calls:
                name = tc.function.name
                try:
                    args = json.loads(tc.function.arguments)
                except json.JSONDecodeError:
                    result = "Invalid JSON arguments"
                else:
                    print(f"\n> {name}({', '.join(f'{k}={repr(v)[:50]}' for k,v in args.items())})")
                    result = self.tools.execute(name, args)
                    # Show truncated result
                    preview = result[:200] + "..." if len(result) > 200 else result
                    print(f"  {preview}")

                self.messages.append({
                    "role": "tool",
                    "tool_call_id": tc.id,
                    "content": result
                })
                self.calls += 1

                # Budget warning
                if self.calls >= TOOL_LIMIT - 5:
                    self.messages[-1]["content"] += f"\n\n⚠️ {TOOL_LIMIT - self.calls} tool calls remaining."

        return self.messages[-1].get("content", "")

# ══════════════════════════════════════════════════════════════════════════════
# CLI / REPL
# ══════════════════════════════════════════════════════════════════════════════

def handle_cmd(cmd: str, agent: Agent) -> bool:
    """Handle slash commands. Returns True if should exit."""
    parts = cmd.split(maxsplit=1)
    c = parts[0].lower()
    
    if c in ["/quit", "/exit", "/q"]:
        return True
    elif c == "/help":
        print("""
Commands:
  /help     Show this help
  /cost     Show token usage
  /clear    Clear conversation
  /model X  Switch model
  /quit     Exit
""")
    elif c == "/cost":
        print(f"Tokens: {agent.tokens:,} | Tool calls: {agent.calls}")
    elif c == "/clear":
        agent.messages = []
        agent.calls = 0
        print("Cleared")
    elif c == "/model":
        if len(parts) > 1:
            agent.model = parts[1]
            print(f"Model: {agent.model}")
        else:
            print(f"Current: {agent.model}")
    else:
        print(f"Unknown: {c}. Type /help")
    return False


def repl(agent: Agent):
    """Interactive REPL."""
    print(f"nano-opencode | {agent.model} | /help for commands\n")
    
    interrupt_count = 0
    def handle_interrupt(sig, frame):
        nonlocal interrupt_count
        interrupt_count += 1
        if interrupt_count > 1:
            print("\nExiting...")
            sys.exit(0)
        print("\n(Ctrl+C again to exit)")
    signal.signal(signal.SIGINT, handle_interrupt)

    while True:
        try:
            interrupt_count = 0
            task = input("nano> ").strip()
            if not task:
                continue
            if task.startswith("/"):
                if handle_cmd(task, agent):
                    break
            else:
                agent.run(task)
                print()
        except EOFError:
            break
        except KeyboardInterrupt:
            continue


def main():
    import argparse
    p = argparse.ArgumentParser(description="nano-opencode: 400 lines to beat them all")
    p.add_argument("task", nargs="*", help="Task to perform (omit for REPL)")
    p.add_argument("-m", "--model", default=MODEL, help="Model to use")
    args = p.parse_args()

    agent = Agent(Path.cwd(), args.model)
    
    if args.task:
        agent.run(" ".join(args.task))
    else:
        repl(agent)


if __name__ == "__main__":
    main()

File: implementations/typescript/nano.ts
#!/usr/bin/env bun
/**
 * nano-opencode: Minimal AI coding agent (~250 LOC)
 * Usage: bun nano.ts "your prompt here"
 *        bun nano.ts  (interactive mode)
 */

import Anthropic from '@anthropic-ai/sdk';
import { readFileSync, writeFileSync, existsSync, readdirSync, statSync } from 'fs';
import { execFileSync } from 'child_process';
import { createInterface } from 'readline';
import { globSync } from 'glob';

// === TOOLS (7 essential) ===
const tools: Anthropic.Tool[] = [
  {
    name: 'read_file',
    description: 'Read file contents',
    input_schema: { type: 'object', properties: { path: { type: 'string' } }, required: ['path'] },
  },
  {
    name: 'write_file',
    description: 'Write content to file',
    input_schema: { type: 'object', properties: { path: { type: 'string' }, content: { type: 'string' } }, required: ['path', 'content'] },
  },
  {
    name: 'edit_file',
    description: 'Replace old_string with new_string in file',
    input_schema: { type: 'object', properties: { path: { type: 'string' }, old_string: { type: 'string' }, new_string: { type: 'string' } }, required: ['path', 'old_string', 'new_string'] },
  },
  {
    name: 'bash',
    description: 'Run shell command (use sh -c for complex commands)',
    input_schema: { type: 'object', properties: { command: { type: 'string' }, args: { type: 'array', items: { type: 'string' } } }, required: ['command'] },
  },
  {
    name: 'glob',
    description: 'Find files matching pattern',
    input_schema: { type: 'object', properties: { pattern: { type: 'string' } }, required: ['pattern'] },
  },
  {
    name: 'grep',
    description: 'Search for pattern in files',
    input_schema: { type: 'object', properties: { pattern: { type: 'string' }, path: { type: 'string' } }, required: ['pattern'] },
  },
  {
    name: 'list_dir',
    description: 'List directory contents',
    input_schema: { type: 'object', properties: { path: { type: 'string' } }, required: ['path'] },
  },
];

// === TOOL EXECUTION ===
function executeTool(name: string, input: Record<string, unknown>): string {
  try {
    switch (name) {
      case 'read_file': {
        const path = input.path as string;
        return existsSync(path) ? readFileSync(path, 'utf-8') : `Error: File not found: ${path}`;
      }

      case 'write_file': {
        const path = input.path as string;
        const content = input.content as string;
        writeFileSync(path, content);
        return `Wrote ${content.length} bytes to ${path}`;
      }

      case 'edit_file': {
        const path = input.path as string;
        if (!existsSync(path)) return `Error: File not found: ${path}`;
        const content = readFileSync(path, 'utf-8');
        const oldStr = input.old_string as string;
        const newStr = input.new_string as string;
        if (!content.includes(oldStr)) return `Error: old_string not found in file`;
        writeFileSync(path, content.replace(oldStr, newStr));
        return `Edited ${path}`;
      }

      case 'bash': {
        const cmd = input.command as string;
        const args = (input.args as string[]) || [];
        // Use sh -c for shell features, or direct command
        const result = args.length > 0
          ? execFileSync(cmd, args, { encoding: 'utf-8', timeout: 30000 })
          : execFileSync('sh', ['-c', cmd], { encoding: 'utf-8', timeout: 30000 });
        return result.slice(0, 50000);
      }

      case 'glob': {
        const pattern = input.pattern as string;
        return globSync(pattern, { cwd: process.cwd() }).slice(0, 100).join('\n') || 'No matches';
      }

      case 'grep': {
        const pattern = input.pattern as string;
        const path = (input.path as string) || '.';
        const result = execFileSync('grep', ['-rn', pattern, path], { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 });
        return result.split('\n').slice(0, 50).join('\n') || 'No matches';
      }

      case 'list_dir': {
        const p = (input.path as string) || '.';
        if (!existsSync(p)) return `Error: Path not found: ${p}`;
        return readdirSync(p).map(f => {
          const stat = statSync(`${p}/${f}`);
          return `${stat.isDirectory() ? 'd' : '-'} ${f}`;
        }).join('\n');
      }

      default:
        return `Unknown tool: ${name}`;
    }
  } catch (e) {
    return `Error: ${e instanceof Error ? e.message : String(e)}`;
  }
}

// === AGENT LOOP ===
const client = new Anthropic();
const messages: Anthropic.MessageParam[] = [];

const SYSTEM = `You are an expert coding assistant. You have tools for reading, writing, editing files, running commands, and searching code. Be concise. Use tools to accomplish tasks.`;

async function chat(userMessage: string): Promise<string> {
  messages.push({ role: 'user', content: userMessage });

  let response = await client.messages.create({
    model: 'claude-sonnet-4-20250514',
    max_tokens: 8192,
    system: SYSTEM,
    tools,
    messages,
  });

  // Agent loop: keep going while there are tool calls
  while (response.stop_reason === 'tool_use') {
    const assistantContent = response.content;
    messages.push({ role: 'assistant', content: assistantContent });

    const toolResults: Anthropic.ToolResultBlockParam[] = [];

    for (const block of assistantContent) {
      if (block.type === 'tool_use') {
        console.log(`\x1b[33m⚡ ${block.name}\x1b[0m`);
        const result = executeTool(block.name, block.input as Record<string, unknown>);
        console.log(`\x1b[90m${result.slice(0, 200)}${result.length > 200 ? '...' : ''}\x1b[0m`);
        toolResults.push({ type: 'tool_result', tool_use_id: block.id, content: result });
      } else if (block.type === 'text' && block.text) {
        process.stdout.write(block.text);
      }
    }

    messages.push({ role: 'user', content: toolResults });

    response = await client.messages.create({
      model: 'claude-sonnet-4-20250514',
      max_tokens: 8192,
      system: SYSTEM,
      tools,
      messages,
    });
  }

  // Extract final text
  const text = response.content
    .filter((b): b is Anthropic.TextBlock => b.type === 'text')
    .map(b => b.text)
    .join('');

  messages.push({ role: 'assistant', content: response.content });
  return text;
}

// === CLI ===
async function main() {
  // Check API key early
  if (!process.env.ANTHROPIC_API_KEY) {
    console.error('Set ANTHROPIC_API_KEY environment variable');
    process.exit(1);
  }

  const args = process.argv.slice(2);

  // Single prompt mode
  if (args.length > 0) {
    const result = await chat(args.join(' '));
    console.log(result);
    return;
  }

  // Interactive mode
  console.log('\x1b[34mnano-opencode\x1b[0m - type your message, /quit to exit\n');

  const rl = createInterface({ input: process.stdin, output: process.stdout });
  const prompt = () => rl.question('\x1b[32m❯\x1b[0m ', async (input) => {
    if (input.trim() === '/quit' || input.trim() === '/exit') {
      rl.close();
      return;
    }
    if (!input.trim()) {
      prompt();
      return;
    }
    try {
      const result = await chat(input);
      console.log(`\n\x1b[34m◆\x1b[0m ${result}\n`);
    } catch (e) {
      console.error(`\x1b[31mError: ${e instanceof Error ? e.message : e}\x1b[0m`);
    }
    prompt();
  });
  prompt();
}

main().catch(console.error);
